api/bazel/BUILD

```diff
@@ -15,3 +18,14 @@ go_proto_compiler(
     valid_archive = False,
     visibility = ["//visibility:public"],
 )
+
+write_json(
+    name = "repository_locations",
+    data = REPOSITORY_LOCATIONS_SPEC,
+)
+
+py_data(
+    name = "repository_locations",
+    filters = ["@envoy//tools/dependency:interpolate_repository_locations"],
```

Comment:
Rather than using a filter style here, can't we just chain two rules, one to first do the interpolation, then another to do the Python encapsulation?

Reply:
we could, but needing a mutation of the data is a common pattern that is repeated, eg for turning extension metadata -> categories, interpolating the versions in the repository data, etc

there are others elsewhere that could do similar (eg in docs generation)

atm i have jq to do this in other places - i love jq but im inclined to remove it as there is not an easy way to bring it in with bazel

whether we do it with jq or python, mutating the data is required a few places, and it feels to me like something that should be included as part of this rule/macro

the way implemented here allows the filters to be chained, which although not of immediate use, provides a pattern i can well imagine using in the future to normalize/cleanup/validate data etc

Reply:
to me the bigger question is whether to add mutation to the write as mentioned above, so the mutated data is cached

Reply:
OK, with https://github.com/envoyproxy/envoy/pull/18086 split out, probably the main discussion on this PR is around this.

I'm still not sure why we can't model this with the Bazel dependency graph and we need to resort to an alternative composition method. Bazel gives you caching for free in this situation.

As I understand it, we have some dictionary D in Bazel land and we want to surface this in a Python file F, after some mutations via Python [M0, M1, .. MN]. This could be a series of Bazel dependencies (assuming just M0 for now).
1. F (created with py_data) takes a dependency on the output of "M0 for D" (an independent target).
2. M0 for D is a  genrule-ish target that takes M0 as a tool dependency and D (possibly as a file?) as a data dependency.

Reply:
what does these transformations ?

Reply:
so in the implemeted schema the pipeline looks like this

`data_source[json/yaml/protobuf?]` -> `filter[pydata]` -> `filter[pydata]` -> `filter[pydata]` -> `py_consumer[pydata]`

the loader/filters are currently untyped but we could use the typing system (eg with `TypedDict` for dicts)

the point is that the data source is source-agnostic, the transformers handle *python* data, and provided the mutated python data to a python lib/consumer

it requires a single rule to register the complete pipeline (altho the filters also need registering as libs).

the schema you are proposing would be more like

`data_source[json/yaml/protobuf?]` -> `filter[pydata]` -> `json? writer` -> `json? loader` -> `filter[pydata]`  -> `json? writer` -> `json? loader` ->  `filter[pydata]`  -> `json? writer` -> `json? loader` ->  `py_consumer[pydata]`

it would require umpteen more rule/registrations, and it forces the data back into being a serialized type in between every mutation

Reply:
I'm not really suggesting that it be this verbose. More that your original pipeline, with 5 steps, be reflected as a series of Bazel graph dependencies. Maybe as a quick sanity check @lizan can take a look here, as he often has some good intuitions on how to model things naturally in Bazel.

Reply:
I want to take a step back here, what kind of data are you trying to provide to python from bazel? I'm questioning about that because there are ways to invoke python tool from bazel (e.g. api_proto_plugin) as either aspect, or rule, it can provide some data from bazel model to python tool. That usually fits bazel model more naturally.

Reply:
i can provide a list of links to the current pending implementations, that make use of it, but there are a few more

i will do it tomorrow